tests/lib/fakeinstaller/main.go

```diff
@@ -0,0 +1,282 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/gadget/install"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/osutil/disks"
+	"github.com/snapcore/snapd/osutil/mkfs"
+)
+
+func firstVol(volumes map[string]*gadget.Volume) *gadget.Volume {
+	for _, vol := range volumes {
+		return vol
+	}
+	return nil
+}
+
+func createPartitions(bootDevice string, volumes map[string]*gadget.Volume) ([]gadget.OnDiskStructure, error) {
+	// TODO: support multiple volumes, see gadget/install/install.go
+	if len(volumes) != 1 {
+		return nil, fmt.Errorf("got unexpected number of volumes %v", len(volumes))
+	}
+
+	diskLayout, err := gadget.OnDiskVolumeFromDevice(bootDevice)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read %v partitions: %v", bootDevice, err)
+	}
+	if len(diskLayout.Structure) > 0 {
+		return nil, fmt.Errorf("cannot yet install on a disk that has partitions")
+	}
+
+	layoutOpts := &gadget.LayoutOptions{
+		IgnoreContent: true,
+	}
+
+	vol := firstVol(volumes)
+	lvol, err := gadget.LayoutVolume(vol, gadget.DefaultConstraints, layoutOpts)
+	if err != nil {
+		return nil, fmt.Errorf("cannot layout volume: %v", err)
+	}
+
+	iconst := &install.CreateOptions{CreateAllMissingPartitions: true}
```

Comment:
(nitpick^9) I would just call that opts or createOpts (but I understand this is beyond nitpicking now, sorry!)